GitUI/BranchTreePanel/RepoObjectsTree.ContextActions.cs

```diff
@@ -37,91 +37,85 @@ partial class RepoObjectsTree : IMenuItemFactory
         /// </summary>
         private MenuItemsGenerator<TagNode> _tagNodeMenuItems;
 
-        private void ContextMenuAddExpandCollapseTree(ContextMenuStrip contextMenu)
+        /// <summary>Toggles the <paramref name="items"/> <see cref="ToolStripItem.Visible"/>
+        /// as well as <see cref="ToolStripItem.Enabled"/> properties depending on <paramref name="enabled"/>.
+        /// Prefer this over only toggling the visibility of an item to enable determining whether the context menu will (once open)
+        /// contain any visible items via <see cref="ToolStripItem.Enabled"/> in <see cref="contextMenu_Opening(object, CancelEventArgs)"/>
+        /// even before the menu itself (as the visual parent) is visble and <see cref="ToolStripItem.Visible"/> of any item therefore false.</summary>
+        private void ToggleMenuItems(bool enabled, params ToolStripItem[] items) => items.ForEach(item => item.Visible = item.Enabled = enabled);
+
+        private void ToggleMenuItems<TNode>(MenuItemsGenerator<TNode> generator, Func<ToolStripItemWithKey, bool> isEnabled) where TNode : class, INode
+            => generator.ForEach(i => ToggleMenuItems(isEnabled(i), i.Item));
+
+        /* add Expand All / Collapse All menu entry
+         * depending on whether node is expanded or collapsed and has child nodes at all */
+        private void ToggleExpandCollapseContextMenu(ContextMenuStrip contextMenu, NodeBase[] selectedNodes)
         {
-            // Add the following to the every participating context menu:
-            //
-            //    ---------
-            //    Collapse All
-            //    Expand All
+            var multiSelectedParents = selectedNodes.HavingChildren().ToArray();
 
-            Tree? treeNode = (contextMenu.SourceControl as TreeView)?.SelectedNode?.Tag as Tree;
-
-            if (contextMenu == menuMain)
+            // selection contains nodes with children
+            if (multiSelectedParents.Length > 0)
```

Comment:
This condition does explicitly apply to the separator only.
The item availability state of the menu items can be set in one statement each.
There is no need for `if {} else {}`.
Applies to the entire file.

Reply:
@mstv I don't understand this either :/ Maybe it becomes clearer if you suggest the change you'd like to see, commented where necessary?

Reply:
This function can be shortened to:
```
        {
            var multiSelectedParents = selectedNodes.HavingChildren().ToArray();
            ToggleMenuItems(multiSelectedParents.Length > 0, tsmiMainMenuSpacer1);
            ToggleMenuItems(multiSelectedParents.Expandable().Any(), mnubtnExpand);
            ToggleMenuItems(multiSelectedParents.Collapsible().Any(), mnubtnCollapse);
        }
```

Reply:
@mstv I think I understand where you're going with this now, thanks for the clarification.
This also relates to above conversations where you suggested that `contextMenu.Items.Clear();` and `AddOnce()` might be unnecessary.
My guess is that this was and implemented like that because the top-level trees share the same context menu `menuMain` containing the Expand/Collapse and Move Up/Down menu items, wheras the `Remotes` tree has its own (`menuRemotes`, see https://github.com/gitextensions/gitextensions/blob/9b7254eb788c7f544a44c153892caddb2ef41a72/GitUI/BranchTreePanel/RepoObjectsTree.cs#L349 ) while sharing the items from `menuMain`. I guess this could be implemented differently, like e.g. the Sort by / order context menu items, which are also shared between multiple menus.
But those diverge from your preferred pattern of adding the context menu items via the designer and instead use a version of the `AddOnce()` pattern. See https://github.com/gitextensions/gitextensions/blob/9b7254eb788c7f544a44c153892caddb2ef41a72/GitUI/BranchTreePanel/RepoObjectsTree.ContextActions.cs#L208-L211

To be clear, I completely embrace your wish to make the context menu behavior more transparent and avoid unexpected side effects. I just don't see how I can fulfill both your wishes to add the menu items from the designer and get rid of the `AddOnce()` pattern without duplicating the items and handlers for the shared entries, which is probably not what you have in mind. Please advise!

Reply:
> to add the menu items from the designer and get rid of the AddOnce() pattern without duplicating the items and handlers for the shared entries, which is probably not what you have in mind.

I was not aware that there are different context menus. Anyway, there is no need to remove and add the fixed menu items every time (except those for user scripts). This should be done once. If the designer does not support sharing, the regarding items can be added in or after `InitializeComponent` once.
Though I think with respect to multi-selections, multiple context menus do not simplify the handling as before with single-selection. That's why there should be one long context menu whose items are enabled/disabled in dependency on the selected items.

Reply:
@mstv Please refer to my answer in above open review conversation explaining why the `AddOnce()` (now `SetLastItem()`) pattern was (and still is) required.

> there should be one long context menu whose items are enabled/disabled in dependency on the selected items

I don't know whether I want to open that can of worms right now if I can avoid it.
How about getting rid of the `SetLastItem()` pattern by adding clones for shared items instead of sharing the same menu item instances across multiple menus? I hope this is a less intrusive change to the context menu than what you suggest. Would that be good enough?

Reply:
Using a single context menu for a single TreeView control will not open a can of worms but will simplify the implementation a lot. A multi-selection can contain any combination of nodes. Having a heuristic which context menu to open in dependency on selected nodes or even on the clicked node makes it difficult to get a consistent behavior.